### PR TITLE
Dpp 1581 consume effective dates and funding bands from win api

### PR DESCRIPTION
**Pull request by chrisfoster76**

### CODE REVIEW

File: src/SFA.DAS.EmployerApprenticeshipsService.Application.UnitTests/Mappers/ApprenticeshipInfoServiceMapperTests/WhenIMapAFramework.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+            //Act
+            var result = _mapper.MapFrom(new List<FrameworkSummary> { CopyOf(_framework) });
+
+            //Assert
+            Assert.IsNotNull(result.Frameworks[0].FundingPeriods);
+            Assert.IsEmpty(result.Frameworks[0].FundingPeriods);
+        }
+
+        private static FrameworkSummary CopyOf(FrameworkSummary source)
```
Comment (lazcool) [inline src/SFA.DAS.EmployerApprenticeshipsService.Application.UnitTests/Mappers/ApprenticeshipInfoServiceMapperTests/WhenIMapAFramework.cs]: There's a TestHelpers.Clone() method for this now.

File: src/SFA.DAS.EmployerApprenticeshipsService.Application.UnitTests/Extensions/ITrainingProgrammeExtensionsTest/WhenDetermingWhetherACourseIsActive.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    public class WhenDetermingWhetherACourseIsActive
+    {
+        [TestCase("2016-01-01", "2016-12-01", "2016-06-01", true, Description = "Within date range")]
+        [TestCase("2016-01-15", "2016-12-15", "2016-01-01", true, Description = "Within date range - ignoring start day")]
+        [TestCase("2016-01-15", "2016-12-15", "2016-12-30", true, Description = "Within date range - ignoring end day")]
+        [TestCase(null, "2016-12-01", "2016-06-01", true, Description="Within date range with no defined course start date")]
+        [TestCase("2016-01-01", null, "2016-06-01", true, Description = "Withing date range, with no defined course end date")]
+        [TestCase(null, null, "2016-06-01", true, Description = "Within date range, with no defined course effective dates")]
+        [TestCase("2016-01-01", "2016-12-01", "2015-06-01", false, Description = "Outside (before) date range")]
+        [TestCase("2016-01-01", "2016-12-01", "2017-06-01", false, Description = "Outside (after) date range")]
```
Comment (lazcool) [inline src/SFA.DAS.EmployerApprenticeshipsService.Application.UnitTests/Extensions/ITrainingProgrammeExtensionsTest/WhenDetermingWhetherACourseIsActive.cs]: I'd be temped to add test cases for either side of both boundaries, i.e.

[TestCase("2016-01-01", "2016-12-01", "2016-01-01", true, Description = "Within date range (first day)")]
[TestCase("2016-01-01", "2016-12-01", "2016-12-31", true, Description = "Within date range (last day)")]
[TestCase("2016-01-01", "2016-12-01", "2015-12-31", false, Description = "Outside (just before) date range")]
[TestCase("2016-01-01", "2016-12-01", "2017-01-01", false, Description = "Outside (just after) date range")]

File: src/SFA.DAS.EmployerApprenticeshipsService.Application.UnitTests/Extensions/ITrainingProgrammeExtensionsTest/WhenDeterminingFundingCap.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                    EffectiveTo = null,
+                    FundingCap = 2000
+                }
+            });
+        }
+
+        [TestCase("2018-05-15", 5000, Description = "Within first funding band")]
+        [TestCase("2018-09-15", 2000, Description = "Within second funding band")]
+        [TestCase("2018-01-01", 0, Description = "Before course start")]
+        [TestCase("2019-06-01", 0, Description = "After course end")]
```
Comment (lazcool) [inline src/SFA.DAS.EmployerApprenticeshipsService.Application.UnitTests/Extensions/ITrainingProgrammeExtensionsTest/WhenDeterminingFundingCap.cs]: Again, it would be useful to have test cases around the boundaries.

File: src/SFA.DAS.EmployerApprenticeshipsService.Application/Services/ApprenticeshipInfoService.cs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        public ApprenticeshipInfoService(ICache cache,
+            IApprenticeshipInfoServiceConfiguration configuration,
+            IApprenticeshipInfoServiceMapper mapper)
+        {
+            _cache = cache;
+            _configuration = configuration;
+            _mapper = mapper;
+        }
+
+        public async Task<StandardsView> GetStandardsAsync(bool refreshCache = false)
```
Comment (lazcool) [inline src/SFA.DAS.EmployerApprenticeshipsService.Application/Services/ApprenticeshipInfoService.cs]: We don't postfix Async :wink:

File: src/SFA.DAS.EmployerApprenticeshipsService.Application/Services/ApprenticeshipInfoService.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        {
+            try
+            {
+                var api = new Providers.Api.Client.ProviderApiClient(_configuration.BaseUrl);
+                var provider = api.Get(ukPrn);
+                return _mapper.MapFrom(provider);
+            }
+            catch (EntityNotFoundException)
+            {
+                return null;
```
Comment (lazcool) [inline src/SFA.DAS.EmployerApprenticeshipsService.Application/Services/ApprenticeshipInfoService.cs]: Would it be better to let the exception propagate?
When we return null, consumers can forget to check for it, e.g. CreateSummary() in EmployerCommitmentsOrchestrator will throw a NullReferenceException if null is returned here.
Of course, the consumers can forget to catch the exception as well, but it's easier to debug a specific exception thrown at the source of the problem, rather than a NullReferenceException at a later point.
Pragmatically though, it might be better if we just check that all the consumers handle a null gracefully.
